Fix #9780 - Blank epJSON inputs needed for ZoneHVAC:EquipmentList

jmarrec opened the pull request:
Pull request overview
---------------------
 - Fixes #9780

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jmarrec committed:
Add a new InputProcessor function: `    const json &getObjectItem(EnergyPlusData &state, std::string_view ObjType, std::string_view ObjName);`

jmarrec committed:
Fix #9780

jmarrec committed:
Test for #9780 - large

jmarrec committed:
Test for #9780 - Extend assertions and minimize greatly the test input

jmarrec commented on tst/EnergyPlus/unit/DataZoneEquipment.unit.cc:
```diff
@@ -155,3 +160,269 @@ TEST_F(EnergyPlusFixture, DataZoneEquipment_TestCalcDesignSpecificationOutdoorAi
     state->dataContaminantBalance->ZoneSysContDemand.deallocate();
 }
+
+TEST_F(EnergyPlusFixture, GetZoneEquipmentData_epJSON)
+{
+
+    state->dataInputProcessing->inputProcessor->epJSON = R"(
+{
+  "ZoneHVAC:EquipmentList": {
+    "Zone1 Equipment List": {
+      "equipment": [
+        {
+          "zone_equipment_cooling_sequence": 1,
+          "zone_equipment_heating_or_no_load_sequence": 1,
+          "zone_equipment_name": "Fan Zone Exhaust 1",
+          "zone_equipment_object_type": "Fan:ZoneExhaust"
+        },
+        {
+          "zone_equipment_cooling_sequence": 3,
+          "zone_equipment_heating_or_no_load_sequence": 2,
+          "zone_equipment_name": "ADU Air Terminal Single Duct Constant Volume No Reheat 1",
+          "zone_equipment_object_type": "ZoneHVAC:AirDistributionUnit"
+        },
+        {
+          "zone_equipment_cooling_sequence": 2,
+          "zone_equipment_heating_or_no_load_sequence": 3,
+          "zone_equipment_name": "ADU Air Terminal Single Duct VAV Reheat 1",
+          "zone_equipment_object_type": "ZoneHVAC:AirDistributionUnit"
+        }
+      ],
```
New unit test showing the error.

jmarrec commented on tst/EnergyPlus/unit/DataZoneEquipment.unit.cc:
```diff
@@ -155,3 +160,269 @@ TEST_F(EnergyPlusFixture, DataZoneEquipment_TestCalcDesignSpecificationOutdoorAi
+    HeatBalanceManager::GetZoneData(*state, ErrorsFound);
+    ASSERT_FALSE(ErrorsFound);
+
+    DataZoneEquipment::GetZoneEquipmentData(*state);
```
It throws here, because it thinks that are only 2 equipments on the ZoneHVAC:EquipmentList

jmarrec commented on tst/EnergyPlus/unit/DataZoneEquipment.unit.cc:
```diff
@@ -155,3 +160,269 @@ TEST_F(EnergyPlusFixture, DataZoneEquipment_TestCalcDesignSpecificationOutdoorAi
+
+    DataZoneEquipment::GetZoneEquipmentData(*state);
+
+    ASSERT_EQ(1, state->dataZoneEquip->ZoneEquipList.size());
+    auto &thisZoneEquipList = state->dataZoneEquip->ZoneEquipList(1);
+    EXPECT_EQ(3, thisZoneEquipList.NumOfEquipTypes);
+
+    EXPECT_TRUE(compare_enums(DataZoneEquipment::LoadDist::Sequential, thisZoneEquipList.LoadDistScheme));
+
+    EXPECT_EQ("FAN ZONE EXHAUST 1", thisZoneEquipList.EquipName(1));
+    EXPECT_EQ("FAN:ZONEEXHAUST", thisZoneEquipList.EquipType(1));
+    EXPECT_TRUE(compare_enums(DataZoneEquipment::ZoneEquip::ZoneExhaustFan, thisZoneEquipList.EquipTypeEnum(1)));
+    EXPECT_EQ(1, thisZoneEquipList.CoolingPriority(1));
+    EXPECT_EQ(1, thisZoneEquipList.HeatingPriority(1));
+    EXPECT_EQ(-1, thisZoneEquipList.SequentialCoolingFractionSchedPtr(1));
+    EXPECT_EQ(-1, thisZoneEquipList.SequentialHeatingFractionSchedPtr(1));
+
+    EXPECT_EQ("ADU AIR TERMINAL SINGLE DUCT CONSTANT VOLUME NO REHEAT 1", thisZoneEquipList.EquipName(2));
+    EXPECT_EQ("ZONEHVAC:AIRDISTRIBUTIONUNIT", thisZoneEquipList.EquipType(2));
+    EXPECT_TRUE(compare_enums(DataZoneEquipment::ZoneEquip::AirDistUnit, thisZoneEquipList.EquipTypeEnum(2)));
+    EXPECT_EQ(3, thisZoneEquipList.CoolingPriority(2));
+    EXPECT_EQ(2, thisZoneEquipList.HeatingPriority(2));
+    EXPECT_EQ(-1, thisZoneEquipList.SequentialCoolingFractionSchedPtr(2));
+    EXPECT_EQ(-1, thisZoneEquipList.SequentialHeatingFractionSchedPtr(2));
+
+    EXPECT_EQ("ADU AIR TERMINAL SINGLE DUCT VAV REHEAT 1", thisZoneEquipList.EquipName(3));
+    EXPECT_EQ("ZONEHVAC:AIRDISTRIBUTIONUNIT", thisZoneEquipList.EquipType(3));
+    EXPECT_TRUE(compare_enums(DataZoneEquipment::ZoneEquip::AirDistUnit, thisZoneEquipList.EquipTypeEnum(3)));
+    EXPECT_EQ(2, thisZoneEquipList.CoolingPriority(3));
+    EXPECT_EQ(3, thisZoneEquipList.HeatingPriority(3));
+    EXPECT_EQ(-1, thisZoneEquipList.SequentialCoolingFractionSchedPtr(3));
+    EXPECT_EQ(-1, thisZoneEquipList.SequentialHeatingFractionSchedPtr(3));
```
Test the rest, might as well.

jmarrec commented on src/EnergyPlus/InputProcessing/InputProcessor.hh:
```diff
@@ -137,6 +137,8 @@ public:
 
     std::pair<std::string, bool> getObjectItemValue(std::string const &field_value, json const &schema_field_obj);
 
+    const json &getObjectItem(EnergyPlusData &state, std::string_view ObjType, std::string_view ObjName);
```
new function in input processor to return a **json** object given it's type and name.

jmarrec commented on src/EnergyPlus/InputProcessing/InputProcessor.cc:
```diff
@@ -898,6 +898,40 @@ void InputProcessor::setObjectItemValue(EnergyPlusData &state,
     }
 }
 
+const json &InputProcessor::getObjectItem(EnergyPlusData &state, std::string_view ObjType, std::string_view ObjName)
+{
+    auto objectInfo = ObjectInfo();
+    objectInfo.objectType = ObjType;
+    objectInfo.objectName = ObjName;
+
+    auto obj_iter = epJSON.find(std::string(ObjType));
+    if (obj_iter == epJSON.end() || obj_iter.value().find(objectInfo.objectName) == obj_iter.value().end()) {
+        auto tmp_umit = caseInsensitiveObjectMap.find(convertToUpper(objectInfo.objectType));
+        if (tmp_umit == caseInsensitiveObjectMap.end()) {
+            // indicates object type not found, see function GeneralRoutines::ValidateComponent
+            ShowFatalError(state, format(R"(ObjectType of type "{}" requested was not found in input)", objectInfo.objectType));
+        }
+        objectInfo.objectType = tmp_umit->second;
+        obj_iter = epJSON.find(objectInfo.objectType);
+    }
+
+    std::string const upperObjName = convertToUpper(objectInfo.objectName);
+
+    for (const auto &[key, val] : obj_iter->items()) {
+        if (convertToUpper(key) == upperObjName) {
+            objectInfo.objectName = key;
+            // markObjectAsUsed(objectInfo.objectType, objectInfo.objectName);
+            auto const find_unused = unusedInputs.find(objectInfo);
+            if (find_unused != unusedInputs.end()) {
+                unusedInputs.erase(find_unused);
+            }
+            return val;
+        }
+    }
+
+    ShowFatalError(state, format(R"(Name "{}" requested was not found in input for ObjectType "{}")", objectInfo.objectType, objectInfo.objectName));
```
implementation

jmarrec commented on src/EnergyPlus/DataZoneEquipment.hh:
```diff
@@ -164,6 +164,9 @@ namespace DataZoneEquipment {
         Num
     };
 
+    constexpr std::array<std::string_view, static_cast<int>(LoadDist::Num)> LoadDistNamesUC = {
+        "SEQUENTIALLOAD", "UNIFORMLOAD", "UNIFORMPLR", "SEQUENTIALUNIFORMPLR"};
```
constexpr array, I plan on using `getEnumerationValue` (before @amirroth mentions it! :))

amirroth replied:
It worked!

jmarrec commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
+        auto &ip = state.dataInputProcessing->inputProcessor;
 
-        ZoneEquipListNum = state.dataInputProcessing->inputProcessor->getObjectItemNum(
-            state, CurrentModuleObject, state.dataZoneEquip->ZoneEquipConfig(ControlledZoneNum).EquipListName);
-        if (ZoneEquipListNum > 0) {
+        ZoneEquipListNum = ip->getObjectItemNum(state, CurrentModuleObject, state.dataZoneEquip->ZoneEquipConfig(ControlledZoneNum).EquipListName);
+
+        if (ZoneEquipListNum <= 0) {
+            ShowSevereError(state,
+                            std::string{RoutineName} + CurrentModuleObject +
+                                " not found = " + state.dataZoneEquip->ZoneEquipConfig(ControlledZoneNum).EquipListName);
+            ShowContinueError(
+                state, "In ZoneHVAC:EquipmentConnections object, for Zone = " + state.dataZoneEquip->ZoneEquipConfig(ControlledZoneNum).ZoneName);
+            state.dataZoneEquip->GetZoneEquipmentDataErrorsFound = true;
+        } else {
+            auto const &epListFields =
+                ip->getObjectItem(state, CurrentModuleObject, state.dataZoneEquip->ZoneEquipConfig(ControlledZoneNum).EquipListName);
```
Here's the use case for the new json InputProcessor::getObjectItem one.

amirroth replied:
May be better to name this function something else, `GetObjectItemJson`? It's essentially a completely different function than the current version.

jmarrec replied:
Renamed to getJSONObjectItem: e3b36fe4a552f0b582794816c0e40871a7961dae

jmarrec committed:
Correct error message

jmarrec commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
-                        ShowContinueError(state, format("only {} in the list.", thisZoneEquipList.NumOfEquipTypes));
-                    state.dataZoneEquip->GetZoneEquipmentDataErrorsFound = true;
-                }
+            std::string loadDistName = ip->getAlphaFieldValue(epListFields, objectSchemaProps, "load_distribution_scheme");
+            thisZoneEquipList.LoadDistScheme = static_cast<DataZoneEquipment::LoadDist>(
+                getEnumerationValue(DataZoneEquipment::LoadDistNamesUC, UtilityRoutines::MakeUPPERCase(loadDistName)));
+            if (thisZoneEquipList.LoadDistScheme == DataZoneEquipment::LoadDist::Invalid) {
+                ShowSevereError(state, format("{}{} = \"{}, Invalid choice\".", RoutineName, CurrentModuleObject, thisZoneEquipList.Name));
+                ShowContinueError(state, format("...load_distribution_scheme=\"{}\".", loadDistName));
+                state.dataZoneEquip->GetZoneEquipmentDataErrorsFound = true;
+            }
```
regular field, with getEnumeration value.

rraustad replied:
This check is not needed. It is not possible for this field to be Invalid. If this field is blank the Json input processor will fill the field with SequentialLoad. If it's not a valid key choice the Json processor will fatal out.

    ZoneHVAC:EquipmentList,
           \extensible:6 Duplicate the six fields Zone Equipment Object Type through Sequential Heating Fraction
           \min-fields 8
      A2 , \field Load Distribution Scheme
           \type choice
           \key SequentialLoad
           \key UniformLoad
           \key UniformPLR
           \key SequentialUniformPLR
           \default SequentialLoad

rraustad replied:
Oh, but this is an extensible object so I am not sure what the Json input processor will do with the extended fields. I would think it would process these extended fields the same way as those that are within min-fields.

mjwitte replied:
> If this field is blank the Json input processor will fill the field with SequentialLoad.

I don't think that's so. `getObjectItem` fills in defaults (up to min-fields), and `getAlphaFieldValue` etc. fills in defaults. Raw epJSON will simply not have a value for this field.

jmarrec commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
-                        ShowContinueError(state,
-                                          "invalid " + cAlphaFields(coolingFractionArrayIdx) + "=[" + AlphArray(coolingFractionArrayIdx) + "].");
-                        ShowContinueError(state, "Schedule does not exist.");
+            auto extensibles = epListFields.find("equipment");
+            if (extensibles != epListFields.end()) {
+                auto extensiblesArray = extensibles.value();
+                thisZoneEquipList.NumOfEquipTypes = static_cast<int>(extensiblesArray.size());
+
+                // Increment overall count of equipment
+                overallEquipCount += thisZoneEquipList.NumOfEquipTypes;
```
Here's the fix. Instead of computing the number of extensible groups based on the entries in the AlphArrays, we use the epJSON extensibles, and the problem goes away

rraustad replied:
I have run into this before where I had to add a schedule name. I thought this was a required field but now I think the problem only happened when there were more that 2 equipment types listed.

jmarrec replied:
Yes, you need 3 to get the original error.

jmarrec commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
+                        state.dataZoneEquip->GetZoneEquipmentDataErrorsFound = true;
+                    }
+
+                    std::string coolingSchName =
+                        ip->getAlphaFieldValue(extensibleInstance, extensionSchemaProps, "zone_equipment_sequential_cooling_fraction_schedule_name");
+                    if (coolingSchName.empty()) {
+                        thisZoneEquipList.SequentialCoolingFractionSchedPtr(ZoneEquipTypeNum) = DataGlobalConstants::ScheduleAlwaysOn;
+                    } else {
+                        thisZoneEquipList.SequentialCoolingFractionSchedPtr(ZoneEquipTypeNum) = GetScheduleIndex(state, coolingSchName);
+                        if (thisZoneEquipList.SequentialCoolingFractionSchedPtr(ZoneEquipTypeNum) == 0) {
+                            ShowSevereError(state, format("{}{} = \"{}\".", RoutineName, CurrentModuleObject, thisZoneEquipList.Name));
+                            ShowContinueError(state,
+                                              format("invalid zone_equipment_sequential_cooling_fraction_schedule_name=[{}].", coolingSchName));
+                            ShowContinueError(state, "Schedule does not exist.");
+                            state.dataZoneEquip->GetZoneEquipmentDataErrorsFound = true;
+                        }
+                    }
```
get the sequential cooling schedule, if provided

jmarrec commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
+                    thisZoneEquipList.CoolingPriority(ZoneEquipTypeNum) =
+                        ip->getIntFieldValue(extensibleInstance, extensionSchemaProps, "zone_equipment_cooling_sequence");
+
+                    if ((thisZoneEquipList.CoolingPriority(ZoneEquipTypeNum) < 0) ||
+                        (thisZoneEquipList.CoolingPriority(ZoneEquipTypeNum) > thisZoneEquipList.NumOfEquipTypes)) {
+                        ShowSevereError(state, format("{}{} = \"{}\".", RoutineName, CurrentModuleObject, thisZoneEquipList.Name));
+                        ShowContinueError(
+                            state, format("invalid zone_equipment_cooling_sequence=[{}].", thisZoneEquipList.CoolingPriority(ZoneEquipTypeNum)));
+                        ShowContinueError(state, "equipment sequence must be > 0 and <= number of equipments in the list.");
+                        if (thisZoneEquipList.CoolingPriority(ZoneEquipTypeNum) > 0) {
+                            ShowContinueError(state, format("only {} in the list.", thisZoneEquipList.NumOfEquipTypes));
+                        }
```
Example getting the cooling priority

Myoldmopar replied:
Yeah, this is nice.  :+1:

rraustad commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
-                    state.dataZoneEquip->GetZoneEquipmentDataErrorsFound = true;
-                }
+            UtilityRoutines::IsNameEmpty(state, thisZoneEquipList.Name, CurrentModuleObject, state.dataZoneEquip->GetZoneEquipmentDataErrorsFound);
 
```
This is a required field, it can't be empty and get this far.

    ZoneHVAC:EquipmentList,
      A1 , \field Name
           \required-field
           \reference ZoneEquipmentLists

amirroth replied:
Translation, you can remove the call to `IsNameEmpty`.

jmarrec replied:
https://github.com/NREL/EnergyPlus/commit/e4a4d843f0b95f06cf46e63769f1173bb212859f

rraustad commented on src/EnergyPlus/DataZoneEquipment.hh:
```diff
@@ -164,6 +164,9 @@ namespace DataZoneEquipment {
 
+    constexpr std::array<std::string_view, static_cast<int>(LoadDist::Num)> LoadDistNamesUC = {
+        "SEQUENTIALLOAD", "UNIFORMLOAD", "UNIFORMPLR", "SEQUENTIALUNIFORMPLR"};
+
```
This should be in the cc since it isn't yet used elsewhere other than DataZoneEquipment.

amirroth replied:
This also worked!

jmarrec replied:
Do you feel strongly about this? Because I find it very awkward (and error-prone) to not have the enum and the std::array that represents it next to each other. 

-----

Just a side note (it may have been done already...), but perhaps it's worth considering switching to another unified type of enum that would allow better handling of both int/value and string. Like the OPENSTUDIO_ENUM macro the SDK has, or something like https://github.com/aantron/better-enums. Here's a compiler explorer example with better enums: https://compiler-explorer.com/z/1GvP4nd7P

```c++
#include <fmt/core.h>

#include "enum.h"


BETTER_ENUM(LoadDist, int, Invalid = -1, Sequential, Uniform, UniformPLR,
            SequentialUniformPLR, Num);

int main() {
    LoadDist c1 = LoadDist::Invalid;
    fmt::print("{}, {}\n", c1, c1._to_string());

    LoadDist c2 = LoadDist::_from_string("Uniform");
    fmt::print("{}, {}\n", c2, c2._to_string());

    LoadDist c3 = LoadDist::_from_string_nocase("UnIFOrMplR");
    fmt::print("{}, {}\n", c3, c3._to_string());

    for (auto c : LoadDist::_values()) {
        fmt::print("{}, {}\n", c, c._to_string());
    }
}
```

amirroth replied:
*Re: `.cc` vs. `.hh`*. Not that strongly. I think if you leave it in the `.hh` you have to declare it `static` or else you will get a copy of the array for each time the `.hh` file is included.

*Re: BETTER_ENUM (or whatever)*. Was not familiar with this. Looks ... interesting. I'm not very good at reading template code, but the implementation looks okay I guess. A faster solution for case-insensitive matching would be nice (e.g., a separate upper-case names array) and looping via recursion is too cute by 10X, but whatever. Are different `BETTER_ENUM` enumerations different types (like `enum class`, i.e., type safe but have to be `static_cast<int>`) or are they all `int` (like `enum`, i.e., not type safe but don't have to be `static_cast<int>`) or are they different types that transparently derive from `int` (i.e., type safe *and* don't have to be `static_cast`<int>` as `enum class` should have been? I will look at `OPENSTUDIO_ENUM` also. Thanks for the references.

jmarrec replied:
I'll move it to the .cc, but for completeness, if it's in a header and declared `static constexpr`, each TU will still get a copy. What we probably want then is `inline constexpr`. As with most C++ things, and especially constexpr things, there's a C++ weekly video about it: https://www.youtube.com/watch?v=QVHwOOrSh3w

As far as the BETTER_ENUM thing, the best way after testing it in action is to ask gcc to expand the macros

```shell
mkdir test_better_enum && cd test_better_enum
wget https://raw.githubusercontent.com/aantron/better-enums/master/enum.h
echo "#include "enum.h" > test.cpp
echo "" >> test.cpp
echo "BETTER_ENUM(LoadDist, int, Invalid = -1, Sequential, Uniform, UniformPLR, SequentialUniformPLR, Num);" >> test.cpp
g++ -std=c++20 -O3 test.cpp -E -o expanded.cpp
```

see https://gist.github.com/jmarrec/8a28cf529742b9c5c603ec77696f86af

amirroth replied:
That's a lot of code per enum (not talking about the included headers, just the stuff that is generated on a per enum basis at the bottom). You are actually getting multiple implementations of the enum, a highly functional and type-safe one in which each enum is implemented as its own class (not its own `enum class`, its own `class`!) and then a more normal and less typesafe one as a C enum. The former is cute, but overkill. The latter is fine if you are willing to give up type safety to the extent that it is even needed for enumerations. FWIW if we were willing to do that, we could replace all of our `enum class` with `enum` and we would be able to drop the `static_cast<int>` everywhere.

This is cute, but I would rather stick with the way we are doing things at least for now. If there is a compelling reason to change, we can do a pass later.

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
+        ZoneEquipListNum = ip->getObjectItemNum(state, CurrentModuleObject, state.dataZoneEquip->ZoneEquipConfig(ControlledZoneNum).EquipListName);
+
+        if (ZoneEquipListNum <= 0) {
+            ShowSevereError(state,
```
`format`

jmarrec replied:
format all the things, done in https://github.com/NREL/EnergyPlus/commit/fb44cc09c8ef7d82f3d61b9c373d5cdf42610385

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
         // for each equipment list
 
         CurrentModuleObject = "ZoneHVAC:EquipmentList";
```
Adding `auto &thisZoneEquipConfig` above should help with readability, etc.

jmarrec replied:
https://github.com/NREL/EnergyPlus/commit/7f949f51228de3e26c5eebac22da9bf003065bee

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
+                // Increment overall count of equipment
+                overallEquipCount += thisZoneEquipList.NumOfEquipTypes;
+
+                thisZoneEquipList.EquipType.allocate(thisZoneEquipList.NumOfEquipTypes);
```
It is not necessary to have both `EquipType` array and `EquipTypeEnum` array. You can always look up the string name from `EquipTypeNames`. PS Once you remove `EquipType`, rename `EquipTypeEnum` to `EquipType`, an `enum` *is* a type.

jmarrec replied:
I'm afraid I won't address this one. I started, but realized that this is isn't doable easily, mostly because there isn't a one-to-one equivalence between the actual string and the ZoneEquip enum, mostly due do this block here:

https://github.com/NREL/EnergyPlus/blob/d7f7f7689e0e2804495811c0c20567f59b5d82f8/src/EnergyPlus/DataZoneEquipment.cc#L561-L572

I'm afraid of breaking things elsewhere, eg here https://github.com/NREL/EnergyPlus/blob/d7f7f7689e0e2804495811c0c20567f59b5d82f8/src/EnergyPlus/DataZoneEquipment.cc#L804-L832

amirroth replied:
Yes. There is a better way of doing this (which I pointed out on another PR), but we will do that separately.

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
+                for (auto &extensibleInstance : extensiblesArray) {
+                    ++ZoneEquipTypeNum;
+
+                    thisZoneEquipList.EquipType(ZoneEquipTypeNum) =
```
Convert to `enum` here.

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
+                    ValidateComponent(state,
+                                      thisZoneEquipList.EquipType(ZoneEquipTypeNum),
+                                      thisZoneEquipList.EquipName(ZoneEquipTypeNum),
+                                      IsNotOK,
```
This parameter sums up my feelings about the `ValidateComponent` function.

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
+
+                thisZoneEquipList.EquipType.allocate(thisZoneEquipList.NumOfEquipTypes);
+                thisZoneEquipList.EquipTypeEnum.allocate(thisZoneEquipList.NumOfEquipTypes);
+                thisZoneEquipList.compPointer.resize(thisZoneEquipList.NumOfEquipTypes + 1);
```
For later maybe, would be better to convert these arrays into a single array of structs.

Myoldmopar replied:
Yeah, it would make the code below a lot easier to clean up as well.

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
+                    std::string heatingSchName =
+                        ip->getAlphaFieldValue(extensibleInstance, extensionSchemaProps, "zone_equipment_sequential_heating_fraction_schedule_name");
+                    if (heatingSchName.empty()) {
+                        thisZoneEquipList.SequentialHeatingFractionSchedPtr(ZoneEquipTypeNum) = DataGlobalConstants::ScheduleAlwaysOn;
```
This `ScheduleAlwaysOn` constant should be moved from `DataGlobalConstants` to `ScheduleManager`. Too many things are in random places.

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
+                        // EquipIndex is used to access specific equipment for a single class of equipment (e.g., PTAC 1, 2 and 3)
+                        thisZoneEquipList.compPointer[ZoneEquipTypeNum] = UnitarySystems::UnitarySys::factory(
+                            state, DataHVACGlobals::UnitarySys_AnyCoilType, thisZoneEquipList.EquipName(ZoneEquipTypeNum), true, 0);
+                        thisZoneEquipList.EquipIndex(ZoneEquipTypeNum) = thisZoneEquipList.compPointer[ZoneEquipTypeNum]->getEquipIndex();
```
Why do you need to save both the pointer to the component and the index to the array that holds the pointer?

rraustad replied:
Because the pointer is used to simulate/access the object and the zone availability manager action for a zone equipment object is stored in an array by component type and component index.

    this->m_AvailStatus = state.dataHVACGlobal->ZoneComp(thisObjectType).ZoneCompAvailMgrs(this->m_EquipCompNum).AvailStatus;

Myoldmopar replied:
I think what this means is that with the right refactoring, we don't need the index anymore, but in the current state, we just aren't there _yet_.

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -382,180 +379,171 @@ void GetZoneEquipmentData(EnergyPlusData &state)
-                        ShowContinueError(state, "Schedule does not exist.");
+            auto extensibles = epListFields.find("equipment");
+            if (extensibles != epListFields.end()) {
+                auto extensiblesArray = extensibles.value();
```
Should this be an `auto &`?

jmarrec replied:
ref not warranted, this returns a const_iterator: https://json.nlohmann.me/api/basic_json/find/

amirroth commented on src/EnergyPlus/InputProcessing/InputProcessor.cc:
```diff
@@ -898,6 +898,40 @@ void InputProcessor::setObjectItemValue(EnergyPlusData &state,
 
+const json &InputProcessor::getObjectItem(EnergyPlusData &state, std::string_view ObjType, std::string_view ObjName)
+{
+    auto objectInfo = ObjectInfo();
```
This is a confusing way to create a local object. `ObjectInfo objectInfo(ObjType, ObjName);` would be clearer, shorter, and probably faster. Also, do you even need this object for anything other than looking for "unused inputs"? And how critical is it to track unused inputs? It seems like something that could add not-insignificant overhead if used everywhere.

jmarrec replied:
I'm copying getObjectItem. 

Also:

* the ctor of ObjectInfo expects a std::string, not a std::string_view,  is `auto objectInfo = ObjectInfo(std::string{ObjType}, std::string{ObjName});` really better?
* Both arguments are of the same type (std::string), so it's kinda easy to mess up the order and not realize you messed up

Unused inputs are reported when DisplayUnusedObjects is enabled in Output:Diagnostics. If not enabled, you still get a warning saying you have unused objects: https://github.com/NREL/EnergyPlus/blob/d7f7f7689e0e2804495811c0c20567f59b5d82f8/src/EnergyPlus/InputProcessing/InputProcessor.cc#L1771-L1775

jmarrec replied:
Done in https://github.com/NREL/EnergyPlus/commit/13c4b06f7b8a4f79c71fa9947d686ae093553be0

jmarrec committed:
Merge branch 'develop' into 9780_epJSON_ZoneHVACEquipmentList

jmarrec committed:
Rename to getJSONObjectItem

jmarrec committed:
Review comment: Take a reference to improve readability

https://github.com/NREL/EnergyPlus/pull/9808#discussion_r1082032634

jmarrec committed:
Review comment: remove IsNameEmpty

https://github.com/NREL/EnergyPlus/pull/9808#discussion_r1082020071

jmarrec committed:
Direct initialize ObjectInfo

https://github.com/NREL/EnergyPlus/pull/9808#discussion_r1082588753

jmarrec committed:
format all the things

jmarrec committed:
Wrong nesting

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into 9780_epJSON_ZoneHVACEquipmentList

jmarrec committed:
Move the constexpr array to the cc file

The `static` is redundant here but I find it clearer and more explicit. 
constexpr implies const and const on namespace scope implies internal linkage

Myoldmopar commented:
Even though this doesn't change the IDD itself, given the input related changes, I think I'll mark this for the IO freeze.

Myoldmopar reviewed (approved):
This seems like a useful addition to the input processor.  Testing locally now.

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -330,20 +332,19 @@ void GetZoneEquipmentData(EnergyPlusData &state)
+                                                         AlphArray(5),
+                                                         state.dataZoneEquip->GetZoneEquipmentDataErrorsFound,
+                                                         DataLoopNode::ConnectionObjectType::ZoneHVACEquipmentConnections,
+                                                         AlphArray(1),
```
Use `thisZoneEquipConfig.ZoneName` if you've already captured it.

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -357,19 +358,19 @@ void GetZoneEquipmentData(EnergyPlusData &state)
         } else {
-            state.dataZoneEquip->ZoneEquipConfig(ControlledZoneNum).ReturnFlowSchedPtrNum = GetScheduleIndex(state, AlphArray(7));
-            if (state.dataZoneEquip->ZoneEquipConfig(ControlledZoneNum).ReturnFlowSchedPtrNum == 0) {
+            thisZoneEquipConfig.ReturnFlowSchedPtrNum = GetScheduleIndex(state, AlphArray(7));
```
`PtrNum` 🤔

amirroth commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -387,197 +388,183 @@ void GetZoneEquipmentData(EnergyPlusData &state)
-                }
+            auto extensibles = epListFields.find("equipment");
+            if (extensibles != epListFields.end()) {
+                auto extensiblesArray = extensibles.value();
```
`.value()` returns an object reference so this needs to be `auto &` or else you are making a local copy.

Myoldmopar commented:
Everything passes happily here, no regressions or anything.  @amirroth I see your comments here, but to get the merge process moving, I'd like to go ahead and merge this in.  I'm happy to continue addressing internal cleanups after I freeze for I/O.

amirroth commented:
There's a bunch of older comments that are more along the line of refactoring suggestions (the usual stuff, e.g., don't use strings for enums, arrays of structs). The only important/immediate comment is that an `auto` should be an `auto &`.

Myoldmopar commented:
I'll fix the `auto &` on a branch today along with the small build warning that crept in.
